mp4/mdat.go

```diff
@@ -84,3 +92,51 @@ func (m *MdatBox) HeaderSize() uint64 {
 func (m *MdatBox) PayloadAbsoluteOffset() uint64 {
 	return m.StartPos + m.HeaderSize()
 }
+
+// setReadSeeker - set readseeker to read Mdat data.
+// When a file is decoded lazily, the Mdat Data byte slice is nil
+// and this readseeker is to read data whenever the data is needed.
+func (m *MdatBox) setReadSeeker(rs io.ReadSeeker) {
+	m.readSeeker = rs
+}
+
+// ReadData reads Mdat data specified by the start and size.
+// Input argument start is the postion relative to the start of a file.
+func (m *MdatBox) ReadData(start, size int64) ([]byte, error) {
+	// The Mdat box was decoded lazily
+	if m.decLazyDataSize > 0 {
+		if m.readSeeker == nil {
+			return nil, errors.New("lazy mdat mode - expects non-nil readseeker to read data")
+		}
+
+		m.mu.Lock()
```

Comment:
This is only reading of data with no simultaneous/interleaved writing, so we should be able to make it lock-free and get rid of this lock.

 I think that setting the readSeeker from the calling thread would be a good way of doing that.